### ISSUE TITLE
Auto-Scale Pie Plots

### ISSUE THREAD

**Issue by alandefreitas** [epezent/implot#195]
It seems like Pie Plots are not scaling properly when the parent plot is not squared. 

<img width="1258" alt="image" src="https://user-images.githubusercontent.com/5369819/111913878-92ce2e00-8a4e-11eb-9bf3-cd9f889d5166.png">

```cpp
if (ImPlot::BeginPlot("##PieCitation", nullptr, nullptr, ImVec2(-1, -1),
                      ImPlotFlags_Equal | ImPlotFlags_NoMousePos,
                      ImPlotAxisFlags_NoDecorations, ImPlotAxisFlags_NoDecorations)) {
    ImPlot::PlotPieChart(labels.data(), pie_sizes.data(), pie_sizes.size(), 0.5, 0.5, 0.4,
                         true, "%.0f");
    ImPlot::EndPlot();
}
```

There is an obvious palliative solution, which is to always define the plot width to be somewhat proportional to the available plot height. However, I might be wrong about that but it seems like doing that has some problems in regards to [intentionality](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rp-what):

- It seems like pie charts are never supposed to be elliptical. In fact, they are usually defined as special functions in a way intended to help users with this usual scaling problem for circles in plots. Some references are the gnuplot [circle](http://gnuplot.sourceforge.net/demo/circles.html) plot type, [matplotlib pie](https://matplotlib.org/stable/gallery/pie_and_polar_charts/pie_and_donut_labels.html#sphx-glr-gallery-pie-and-polar-charts-pie-and-donut-labels-py) function, and really any plotting library I can think of. It's actually very hard to think of a use case for elliptical pie plots. You can see how the pies are independent from the parent plot in these libraries:

![](http://gnuplot.sourceforge.net/demo/circles.1.png)

![](https://matplotlib.org/stable/_images/sphx_glr_pie_and_donut_labels_001.png)

- This might be my lack of ability with ImGui, but it seems like it's not always easy to manually adjust this width/height ratio because these are often defined as (-1,-1) to take the whole area available. This area usually doesn't match the window width/height in equal proportions. But this might be just my ignorance about ImGui. There might be a simple way to do that.
- Squaring the plot has a limitation that the labels will tend to be on top of the pie. In this case, we can compare [the matplotlib pies](https://github.com/epezent/implot/issues/new) with the [Implot pies](https://raw.githubusercontent.com/wiki/epezent/implot/screenshots3/pie.gif).

![](https://matplotlib.org/stable/_images/sphx_glr_pie_and_donut_labels_001.png)

![](https://raw.githubusercontent.com/wiki/epezent/implot/screenshots3/pie.gif)

Thanks again for the great library 😀. I hope these comments are useful.

**Comment by epezent** [comment]
Yes Pie charts needs some work. They were hastily added a while back when we had just gotten started, but no one has raised an issue with them yet. This was before we had ImPlotFlags_Equal, hence the two not working together as expected. I can take a look sometime soon.

**Comment by epezent** [comment]
@alandefreitas , sorry for such a long delay in handling this. I've tested and thought about this a bit more and:

1) Pie plots are always circular when `ImPlotFlags_Equal` is used. So, if users want to ensure circularness they should just enable this flag.
2) matplotlib will in fact produce an elliptical pie plot if axis scaling is not equal:

```python
import matplotlib.pyplot as plt
labels = 'Frogs', 'Hogs', 'Dogs', 'Logs'
sizes = [15, 30, 45, 10]
fig1, ax1 = plt.subplots()
ax1.pie(sizes, labels=labels, autopct='%1.1f%%',)
ax1.axis('auto') # 'equal' makes the pie plot circular
plt.show()
```

![Figure_1](https://user-images.githubusercontent.com/29577475/121292201-692ee100-c89e-11eb-9987-82d1d0c1d4d5.png)

So, from my perspective ImPlot and matplotlib are behaving the same way in regard to pie plots. Therefore, I do not feel it's necessary to change the behavior of pie plots at this time. Closing this issue for now.

**Comment by alandefreitas** [comment]
@epezent, thanks for having a look.

You're correct. I didn't notice the [matplotlib example](https://matplotlib.org/stable/gallery/pie_and_polar_charts/pie_features.html) used `ax1.axis('equal')` and `ImPlotFlags_Equal` worked for pie plots.

My suspicion is that what's happening is `ImPlotFlags_Equal` only works when the plot size is not `ImVec2(-1, -1)` though.

I have this piece of code with `ImPlotFlags_Equal` somewhere in an application (with ImPlot 0.9):

```cpp
        // std::vector<char *> labels = ...; std::vector<float> pie_sizes = ...;
        if (ImPlot::BeginPlot("##PieCitation", nullptr, nullptr, ImVec2(-1, -1),
                              ImPlotFlags_Equal | ImPlotFlags_NoMousePos,
                              ImPlotAxisFlags_NoDecorations, ImPlotAxisFlags_NoDecorations)) {
            ImPlot::PlotPieChart(labels.data(), pie_sizes.data(), pie_sizes.size(), 0.5, 0.5, 0.4,
                                 true, "%.0f");
            ImPlot::EndPlot();
        }
```

and it renders this pie:

![image](https://user-images.githubusercontent.com/5369819/121404419-5239ce80-c932-11eb-981e-c416162a9fc3.png)

The implot demo uses:

```cpp
ImPlot::SetNextPlotLimits(0,1,0,1,ImGuiCond_Always);        
if (ImPlot::BeginPlot("##Pie1", NULL, NULL, ImVec2(250,250), ImPlotFlags_Equal | ImPlotFlags_NoMousePos, ImPlotAxisFlags_NoDecorations, ImPlotAxisFlags_NoDecorations)) {            
    ImPlot::PlotPieChart(labels1, data1, 4, 0.5, 0.5, 0.4, normalize, "%.2f");            
    ImPlot::EndPlot();        
}
```

and it renders the pie with the proper scales.

So it seems like what's happening is `ImPlotFlags_Equal` isn't working with the plot size is `ImVec2(-1, -1)`. I might just be doing something wrong but that what it looks like.

**Comment by epezent** [comment]
I can't reproduce this issue. Are you on the latest commit?

```cpp
ImGui::Begin("My Window");
static const char* labels1[]   = {"Frogs","Hogs","Dogs","Logs"};
static float data1[]           = {0.15f,  0.30f,  0.2f, 0.05f};
if (ImPlot::BeginPlot("##Pie1", NULL, NULL, ImVec2(-1,-1), ImPlotFlags_Equal)) {
    ImPlot::PlotPieChart(labels1, data1, 4, 0.5, 0.5, 0.4);
    ImPlot::EndPlot();
}
ImGui::End();
```

![pie](https://user-images.githubusercontent.com/29577475/121407624-5fe05b80-c914-11eb-8b44-8ecae41fff31.gif)

**Comment by epezent** [comment]
I did however notice that pie plots weren't setup to respond to fitting. I've pushed a fix for this. Now, with all the flags you have enabled plus `ImPlotAxisFlags_AutoFit` on both y and x, I think you get a pretty decent pie plot experience:

![pie2](https://user-images.githubusercontent.com/29577475/121408984-d6318d80-c915-11eb-98fe-04b16e778092.gif)

**Comment by alandefreitas** [comment]
I was using v0.9. I downloaded the latest commit now.

The pie plot initially looked the same, but I commented out that previous `ImPlot::SetNextPlotLimits(0, 1, 0, 1, ImGuiCond_Always);` and it looks good now 😀.

![image](https://user-images.githubusercontent.com/5369819/121418966-d9427300-c941-11eb-895f-e7333d251e65.png)

with

```cpp
        // ImPlot::SetNextPlotLimits(0, 1, 0, 1, ImGuiCond_Always);
        if (ImPlot::BeginPlot("##PieCitation", nullptr, nullptr, ImVec2(-1, -1),
                              ImPlotFlags_Equal | ImPlotFlags_NoMousePos,
                              ImPlotAxisFlags_NoDecorations, ImPlotAxisFlags_NoDecorations)) {
            ImPlot::PlotPieChart(labels.data(), pie_sizes.data(), static_cast<int>(pie_sizes.size()), 0.5, 0.5, 0.4,
                                 true, "%.0f");
            ImPlot::EndPlot();
        }
```

`SetNextPlotLimits` is a command I had been using for almost all other plots, as I needed the limits to adjust to the data, whose ranges vary a lot. I expected that to make sense in pie plots, where the axis ranges are not even in my control, but it seems like it makes `ImPlotFlags_Equal` have no effect on the plot. Maybe `SetNextPlotLimits` just doesn't make sense for pie plots.

I don't know if that's expected behavior but it's an edge case anyway. Probably not expected behavior, as this reads to me as the user is asking (i) create a plot in a way that best adjusts to the data (`SetNextPlotLimits(0,1,0,1)`), (ii) make the axis scales proportional (`ImPlotFlags_Equal`), and (iii) give preference to `ImPlotFlags_Equal` over `SetNextPlotLimits` if they are in conflict (it doesn't seem like they are in conflict though). 

I moved back to v0.9 and it works fine even there! 🥳 It actually looks a little better:

![image](https://user-images.githubusercontent.com/5369819/121420943-0728b700-c944-11eb-8ee5-005b50e49cd3.png)

🎉
